File: 472. Concatenated Words/main.py
from typing import List, Set

class Solution:
    def isConcat(self, target: str, uset: Set) -> bool:
        if len(uset) == 0:
            return False
        dp = [True] + [False] * len(target)
        for i in range(1,len(target)+1):
            for j in range(i):
                if dp[j] and (target[j:i] in uset):
                    dp[i] = True
                    break
        return dp[-1]

    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        uset = set(words)
        ans = []
        for word in words:
            if word=="":
                continue
            uset.remove(word)
            if self.isConcat(word, uset):
                ans.append(word)
            uset.add(word)

        return ans


if __name__ == "__main__":
    words = ["cat","cats","catsdogcats","dog","dogcatsdog","hippopotamuses","rat","ratcatdogcat"]
    
    sol = Solution()
    ans = sol.findAllConcatenatedWordsInADict(words)
    for w in ans:
        print(w)

File: 472. Concatenated Words/main.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <string>

using namespace std;

class Solution {
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        unordered_set<string> uset(words.begin(), words.end());
        vector<string> ans;
        for(string word : words) {
            if(word.length()==0)
                continue;
            uset.erase(word);
            vector<bool> dp(word.length()+1, false);
            dp[0] = true;
            for(int i=1; i<=word.length(); i++) {
                for(int j=0; j<i; j++) {
                    if(dp[j] && uset.count(word.substr(j,i-j))) {
                        dp[i] = true;
                        break;
                    }
                }
            }
            uset.insert(word);

            if(dp[word.length()]) {
                ans.push_back(word);
            }
        }
        return ans;
    }
};

int main()
{
    vector<string> words = {
        "cat","cats","catsdogcats","dog","dogcatsdog","hippopotamuses","rat","ratcatdogcat"
    };

    Solution sol;
    vector<string> ans = sol.findAllConcatenatedWordsInADict(words);
    for(string s : ans) {
        cout<<s<<endl;
    }

    return 0;
}